### PR TITLE
Numpy v2 support

### PR DESCRIPTION
**Pull request by pimvenderbosch**
Numpy made a v2 release quite some time ago. We should support v2 while keeping backwards compatibility with v1.

### CODE REVIEW

**Review by HGSilveri — COMMENTED** [review]
Looking good, thanks @pimvenderbosch !

File: pulser-core/pulser/math/abstract_array.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def detach(self) -> AbstractArray:
             return AbstractArray(cast(torch.Tensor, self._array).detach())
         return self
 
-    def __array__(self, dtype: Any = None) -> np.ndarray:
+    def __array__(self, dtype: Any = None, copy: Any = None) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (pimvenderbosch) [reply pulser-core/pulser/math/abstract_array.py:151]: I guess we should, but we can't pass it to `np.__array__` because of backwards compatibility

**Comment by pimvenderbosch** [comment]
The tests following tests fail:
- `tests/simulation/test_simulation::test_noise`
- `tests/simulation/test_simulation::test_noise_xy`
 I did a deep dive to find the reason for this. When installing a different version of numpy, a different qutip version is installed. This is the reason that even though setting the seed the same, and this is guaranteed to be reproducible, it still evaluates differently. In between qutip versions there must have been a change in how they generate randomness...

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -9,6 +9,13 @@ repos:
     hooks:
       - id: flake8
 
+  # Ruff only checks numpy v2 compatibility at this moment
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.11.1
+    hooks:
+      - id: ruff
+        args: [ --fix ]
```
[the remark this answers is not in the document]
Comment (pimvenderbosch) [reply .pre-commit-config.yaml:17]: Not sure. I thought this would ensure that we don't push numpy v2 incompatible code.

File: pulser-core/pulser/math/abstract_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:151]: Is it just supposed to make a copy if `True`? If so, we could do this:
```suggestion
    def __array__(self, dtype: Any = None, copy: Any = None) -> np.ndarray:
        _self = self.copy() if copy else self
        ...
```

**Comment by HGSilveri** [comment]
> * I did a deep dive to find the reason for this. When installing a different version of numpy, a different qutip version is installed. This is the reason that even though setting the seed the same, and this is guaranteed to be reproducible, it still evaluates differently. In between qutip versions there must have been a change in how they generate randomness...

Ouf, that's prickly... If it's just a matter of `qutip` random number generation, I'd say that this warrants skipping the test conditionally on the installed version of `qutip` (e.g. with [skipif](https://docs.pytest.org/en/stable/how-to/skipping.html#id1))

File: .pre-commit-config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (HGSilveri) [reply .pre-commit-config.yaml:17]: Not everyone has `pre-commit` activated (in fact, most people don't judging by the amount of times CI fails on things pre-commit would catch).

[the remark this answers is not in the document]
Comment (pimvenderbosch) [reply .pre-commit-config.yaml:17]: ow, I just assumed that the CI would run the pre-commits. But if not then, yes should be added to CI

File: pulser-core/pulser/math/abstract_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pimvenderbosch) [reply pulser-core/pulser/math/abstract_array.py:151]: Yes indeed. As per the [migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html):

```
 If the keywords are added to the __array__ signature, then for:

-  copy=True and any dtype value always return a new copy,
-  copy=None create a copy if required (for example by dtype),
-  copy=False a copy must never be made. If a copy is needed to return a numpy array or satisfy dtype, then raise an exception (ValueError).
```

File: .github/workflows/pulser-setup/action.yml [hunk 3]
```diff
@@ -13,6 +13,10 @@ inputs:
     description: Whether to include pytorch
     required: false
     default: "with-torch"
+  numpy:
+    decription: Which version of numpy to install
```
Comment (HGSilveri) [inline .github/workflows/pulser-setup/action.yml:17]:
```suggestion
    description: Which version of numpy to install
```

File: .github/workflows/pulser-setup/action.yml [hunk 4]
```diff
@@ -33,6 +37,11 @@ runs:
       run: |
         python -m pip install --upgrade pip
         make dev-install-no-torch
+    - name: Install numpy v1
+      if: ${{ inputs.numpy == 'numpy-v1'}}
+      shell: bash
+      run: |
+        pip install numpy~=1.22
```
Comment (HGSilveri) [inline .github/workflows/pulser-setup/action.yml:44]: Shouldn't we pin to the last minor of numpy 1?

File: pulser-core/pulser/math/__init__.py [hunk 5]
```diff
@@ -16,7 +16,7 @@
 from __future__ import annotations
 
 from collections.abc import Sequence
-from typing import cast, Protocol, TypeVar
+from typing import cast, Protocol, TypeVar, Any
```
Comment (HGSilveri) [inline pulser-core/pulser/math/__init__.py:19]: I'm surprised that isort does not complain about this...

**Review by HGSilveri — APPROVED** [review]
LGTM, thanks @pimvenderbosch !

**Comment by HGSilveri** [comment]
Actually, one last thing: why don't we see the numpy version tests in the CI?

**Comment by pimvenderbosch** [comment]
> Actually, one last thing: why don't we see the numpy version tests in the CI?

Because I only enabled v1 tests on PR, but a matrix v1, v2 on develop/master. Better to enable the matrix on PR as well, I guess.

**Comment by pimvenderbosch** [comment]
> > Actually, one last thing: why don't we see the numpy version tests in the CI?
> 
> Because I only enabled v1 tests on PR, but a matrix v1, v2 on develop/master. Better to enable the matrix on PR as well, I guess.

Added this now. Not sure if this is desirable because the number of tests is maybe a bit much now and can slow down the pipeline. What do you think?

**Comment by HGSilveri** [comment]
> > 
> 
> Added this now. Not sure if this is desirable because the number of tests is maybe a bit much now and can slow down the pipeline. What do you think?

I think we only need to vary the numpy version on the `with-torch` tests, doing it for the `no-torch` too sounds like overkill indeed.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -18,6 +18,7 @@ jobs:
         # Python 3.9 does not run on macos-latest (14)
         # Uses macos-13 for 3.9 and macos-latest for >=3.10 
         os: [ubuntu-latest, macos-13, macos-latest, windows-latest]
+        numpy: ["numpy-v1", "numpy-v2"]
```
Comment (HGSilveri) [inline .github/workflows/test.yml:21]: Before we merge, we should check that the tests pass on all OSs. What I normally do for that is make a commit where this CI job runs on pull request and if everything passes I then revert it

**Comment by pimvenderbosch** [comment]
Turns out the latest torch version available on MacOS 13 and python 3.9 is incompatible with numpy v2. Users can use numpy v2 instead or upgrading python >3.9.

File: pulser-core/requirements.txt [hunk 7]
```diff
@@ -3,5 +3,5 @@ referencing
 matplotlib < 4
 packaging   # This is already required by matplotlib but we use it too
 # Numpy 1.20 introduces type hints, 1.24.0 breaks matplotlib < 3.6.1
-numpy >= 1.20, != 1.24.0, < 2
+numpy >= 1.20, != 1.24.0
```
[the remark this answers is not in the document]
Comment (pimvenderbosch) [reply pulser-core/requirements.txt:6]: It does work with python3.9, the problem above was caused by the torch version being available on 3.9 **and** the os version. So think this is too restrictive.
